### ISSUE TITLE
API `getSemanticDiagnostics` reports wrong error for code extends interface in namespace

### ISSUE THREAD

**Issue by plantain-00** [microsoft/TypeScript#55729]
### 🔎 Search Terms

"getSemanticDiagnostics namespace"

### 🕗 Version & Regression Information

- I was unable to test this on prior versions because _______

### 💻 Code

```ts
namespace A {
  export interface C {}
}

interface B extends A.C {
}
```

### 🙁 Actual behavior

```
    messageText: "Cannot extend an interface 'A.C'. Did you mean 'implements'?",
    category: 1,
    code: 2689,
    reportsUnnecessary: undefined,
    reportsDeprecated: undefined
```

### 🙂 Expected behavior

No error.

**Comment by RyanCavanaugh** [comment]
We need a full repro of how you're invoking the TS API

**Comment by plantain-00** [comment]
@RyanCavanaugh 

`package.json`

```json
{
  "dependencies": {
    "typescript": "^5.2.2"
  }
}
```

`script.js`

```js
const ts = require('typescript')

const program = ts.createProgram(['./target.ts'], {})
const checker = program.getTypeChecker()
for (const sourceFile of program.getSourceFiles()) {
  if (sourceFile.fileName.includes('node_modules')) continue
  const visitor = (node) => {
    if (ts.isIdentifier(node)) {
      checker.getTypeAtLocation(node)
    }
    return ts.forEachChild(node, (node) => visitor(node))
  }
  visitor(sourceFile)
  console.info(sourceFile.fileName, program.getSemanticDiagnostics(sourceFile))
}
```

`target.ts`

```ts
namespace A {
  export interface C { }
}

interface B extends A.C {
}
```

**Comment by jakebailey** [comment]
I assume that if you move the `getSemanticDiagnostics` call before the visitor, the problem doesn't show up? I'm not totally sure how well things work when you're mixing and matching those.

**Comment by plantain-00** [comment]
> I assume that if you move the `getSemanticDiagnostics` call before the visitor, the problem doesn't show up? I'm not totally sure how well things work when you're mixing and matching those.

Yes.

**Comment by jakebailey** [comment]
To be clear, I'm not saying that this _isn't_ a bug, just taking a guess as to what's causing it. 😄
